File: app01_java_block/src/test/java/com/gandarych/legalchain/consensus/ConsensusStrategyTest.java
package com.gandarych.legalchain.consensus;

import com.gandarych.legalchain.core.Block;
import com.gandarych.legalchain.core.BlockCandidate;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;

/** Each strategy must seal a candidate and re-validate its own proof deterministically. */
class ConsensusStrategyTest {

    private final BlockCandidate candidate = new BlockCandidate(
            1, 1_700_000_000_000L, "a".repeat(64), List.of(), "validator-x");

    @Test
    void proofOfWorkMeetsDifficultyAndValidates() {
        ProofOfWorkStrategy pow = new ProofOfWorkStrategy();
        Block block = pow.seal(candidate);

        assertThat(block.hash()).startsWith(ProofOfWorkStrategy.DIFFICULTY_PREFIX);
        assertThat(pow.validate(block)).isTrue();
    }

    @Test
    void proofOfStakeSelectionIsDeterministic() {
        ProofOfStakeStrategy pos = new ProofOfStakeStrategy();
        Block first = pos.seal(candidate);
        Block second = pos.seal(candidate);

        assertThat(first.proof()).isEqualTo(second.proof()); // same seed, same winner
        assertThat(pos.validate(first)).isTrue();
    }

    @Test
    void bftRecordsVerifiableQuorum() {
        BftStrategy bft = new BftStrategy();
        Block block = bft.seal(candidate);

        assertThat(block.proof()).startsWith("quorum=4/4");
        assertThat(bft.validate(block)).isTrue();
    }

    @Test
    void strategiesRejectForeignProofs() {
        ProofOfStakeStrategy pos = new ProofOfStakeStrategy();
        Block posBlock = pos.seal(candidate);

        assertThat(new BftStrategy().validate(posBlock)).isFalse();
    }
}

File: app02_cpp_block/tests/ConsensusStrategyTest.cpp
// Each strategy must seal a candidate and re-validate its own proof
// deterministically. Mirrors app01_java_block's ConsensusStrategyTest.java.
#include <gtest/gtest.h>

#include "consensus/BftStrategy.h"
#include "consensus/ProofOfStakeStrategy.h"
#include "consensus/ProofOfWorkStrategy.h"
#include "core/BlockCandidate.h"

using namespace legalchain;

namespace {
core::BlockCandidate candidate() {
    return core::BlockCandidate(1, 1700000000000LL, std::string(64, 'a'), {}, "validator-x");
}
} // namespace

TEST(ConsensusStrategyTest, ProofOfWorkMeetsDifficultyAndValidates) {
    consensus::ProofOfWorkStrategy pow;
    core::Block block = pow.seal(candidate());

    EXPECT_EQ(block.hash().compare(0, 4, consensus::ProofOfWorkStrategy::DIFFICULTY_PREFIX), 0);
    EXPECT_TRUE(pow.validate(block));
}

TEST(ConsensusStrategyTest, ProofOfStakeSelectionIsDeterministic) {
    consensus::ProofOfStakeStrategy pos;
    core::Block first = pos.seal(candidate());
    core::Block second = pos.seal(candidate());

    EXPECT_EQ(first.proof(), second.proof()); // same seed, same winner
    EXPECT_TRUE(pos.validate(first));
}

TEST(ConsensusStrategyTest, BftRecordsVerifiableQuorum) {
    consensus::BftStrategy bft;
    core::Block block = bft.seal(candidate());

    EXPECT_EQ(block.proof().compare(0, 10, "quorum=4/4"), 0);
    EXPECT_TRUE(bft.validate(block));
}

TEST(ConsensusStrategyTest, StrategiesRejectForeignProofs) {
    consensus::ProofOfStakeStrategy pos;
    core::Block posBlock = pos.seal(candidate());

    consensus::BftStrategy bft;
    EXPECT_FALSE(bft.validate(posBlock));
}
